nj2k/src/org/jetbrains/kotlin/nj2k/conversions/ContinueStatementConversion.kt

```diff
@@ -0,0 +1,151 @@
+            return null
+        }
+        return JKLabeledExpression(
+            JKWhileStatement(
+                loopStatement.condition.detached(loopStatement),
+                body
+            ),
+            listOf(loop.copyTreeAndDetach())
+        ).asStatement()
+    }
+
+    private fun convertSyntheticWhileStatement(loopStatement: JKKtConvertedFromForLoopSyntheticWhileStatement): JKStatement {
+        var needLabel = false
+        val continueStatementConverter = object : RecursiveApplicableConversionBase(context) {
+            override fun applyToElement(element: JKTreeElement): JKTreeElement {
+                if (element !is JKContinueStatement) return recurse(element)
+                val elementPsi = element.psi<PsiContinueStatement>()!!
+                if (elementPsi.findContinuedStatement()?.toContinuedLoop() != loopStatement.whileStatement.psi<PsiForStatement>()) {
+                    return recurse(element)
+                }
+                if (element.parentIsWhenCase() && element.label is JKLabelEmpty) {
+                    element.label = JKLabelText(loop.copyTreeAndDetach())
+                    needLabel = true
+                }
+                val statements = loopStatement.forLoopUpdaters.map { it.copyTreeAndDetach() } + element.copyTreeAndDetach()
+                return if (element.parent is JKBlock)
+                    JKBlockStatementWithoutBrackets(statements)
+                else JKBlockStatement(JKBlockImpl(statements))
+            }
+        }
+        val body = continueStatementConverter.applyToElement(loopStatement.whileStatement::body.detached())
+        if (needLabel) {
+            return JKBlockStatementWithoutBrackets(
+                listOf(
+                    loopStatement::variableDeclaration.detached(),
+                    JKLabeledExpression(
+                        JKWhileStatement(
+                            loopStatement.whileStatement::condition.detached(),
+                            body as JKStatement
+                        ),
+                        listOf(loop.copyTreeAndDetach())
+                    ).asStatement()
+                )
+            )
+        }
+        return JKKtConvertedFromForLoopSyntheticWhileStatement(
+            loopStatement::variableDeclaration.detached(),
+            JKWhileStatement(
+                loopStatement.whileStatement::condition.detached(),
+                body as JKStatement
+            )
+        )
+    }
+}
+
+private class ContinueStatementConverter(loopStatement: JKStatement, context: NewJ2kConverterContext) {
+    private val recursiveApplicableConversion = object : RecursiveApplicableConversionBase(context) {
+        override fun applyToElement(element: JKTreeElement): JKTreeElement {
+            if (element !is JKContinueStatement) return recurse(element)
+            val elementPsi = element.psi<PsiContinueStatement>()!!
+            val loopPsi = loopStatement.psi<PsiWhileStatement>()
```

Comment:
Instead of checking for 3 different types of PSI loops we you can just use their common interface `PsiLoopStatement` or just not specify a type argument at all (there is a `psi` property for that)